main.go

```diff
@@ -109,54 +117,137 @@ func runCmdFunc(cmd *cobra.Command, args []string) error {
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// processFiles takes a list of files, and for each, attempts to convert it
+// to a JSON value and runs the jq program it
+func processFiles(paths []string, program string, flags flags) error {
+	for _, path := range paths {
+		fileInfo, err := readFile(path, flags)
+		if err != nil {
+			return err
+		}
+
+		err = runJQ(program, fileInfo.data, fileInfo.encoder, flags)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// slurpFiles takes a list of files, and for each, attempts to convert it to
+// a JSON value and appends each JSON value to an array, and passes that array
+// as the input to the jq program.
+func slurpFiles(paths []string, program string, encoder formats.Encoding, flags flags) error {
+	// we ignore the errors because byte.Buffers generally do not return
+	// error on write, and instead only panic when they cannot grow the
+	// underlying slice.
+	var buf bytes.Buffer
+
+	// append the first array bracket
+	buf.WriteRune('[')
+
+	// iterate over each file, appending it's contents to an array
+	for i, path := range paths {
+		fileInfo, err := readFile(path, flags)
 		if err != nil {
 			return err
 		}
+		// only handle files with content
+		if len(bytes.TrimSpace(fileInfo.data)) != 0 {
+			buf.Write(fileInfo.data)
+			// append the comma if it isn't the last item in the array
+			if i != len(paths)-1 {
+				buf.WriteRune(',')
+			}
+		}
+	}
+	// apend the last array bracket
+	buf.WriteRune(']')
+
+	data := buf.Bytes()
+	if len(data) == 0 {
+		return nil
+	}
+
+	err := runJQ(program, data, encoder, flags)
+	if err != nil {
+		return err
 	}
 
 	return nil
 }
 
-func processPathArg(pathArg, program string, flags flags) error {
-	path := os.ExpandEnv(pathArg)
+type fileInfo struct {
+	path             string
+	data             []byte
+	encoder, decoder formats.Encoding
+}
+
+func readFile(path string, flags flags) (*fileInfo, error) {
+	path = os.ExpandEnv(path)
 	fileBytes, err := ioutil.ReadFile(path)
 	if err != nil {
-		return fmt.Errorf("failed to read file at %s: `%s`", path, err)
+		return nil, fmt.Errorf("failed to read file at %s: `%s`", path, err)
 	}
 
 	// If there was no input, there's no output!
 	if len(fileBytes) == 0 {
-		return nil
+		return nil, nil
 	}
 
 	decoder, err := determineDecoder(flags.inputFormat, path, fileBytes)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
 	encoder, err := determineEncoder(flags.outputFormat, decoder)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	jsonifiedFile, err := decoder.MarshalJSONBytes(fileBytes)
+	data, err := decoder.MarshalJSONBytes(fileBytes)
 	if err != nil {
-		return fmt.Errorf("failed to jsonify file at %s: `%s`", path, err)
+		return nil, fmt.Errorf("failed to jsonify file at %s: `%s`", path, err)
 	}
 
-	resultJvs, err := execJQProgram(program, path, jsonifiedFile)
+	return &fileInfo{path: path, data: data, encoder: encoder, decoder: decoder}, nil
```

Comment:
Remove the names to the fields here and the compiler will enforce that you provide all of them.

Reply:
But then ordering can change which will break it later, and also you could mix encoder/decoder. Thoughts?

Reply:
Resolving because now we're not assigning to all values in this struct. LMK if you think we should do something similar int he new code.